### ISSUE TITLE
revisit zson scanner and add more comprehensive tests

### ISSUE THREAD

**Issue by mccanne** [brimsec/zq#2057]
Since changing the ZSON parser to stream its input, we've run into a couple bugs.  This change was hurriedly put in so we should go back and study the code more completely and write more extensive tests that stress the streaming nature of the zson scanner.

**Comment by philrz** [comment]
Repro is with `zq` commit 79b4821.

As part of the TZNG->ZSON test conversions, I'm working with this test:

[https://github.com/brimsec/zq/blob/master/ppl/zqd/handlers_test.go#L578-L591](https://github.com/brimsec/zq/blob/master/ppl/zqd/handlers_test.go#L578-L591)

Run manually in its original TZNG form, it gives an error message as expected:

```
$ cat in.tzng 
#0:record[_path:string,ts:time,uid:bstring
0:[conn;1;CBrzd94qfowOqJwCHa;]

$ zq -z in.tzng 
in.tzng: format detection error
	tzng: line 1: syntax error parsing type string
...
```

But if I try to construct similar malformed ZSON variations by dropping bits of input, it silently eats them without error.

```
$ cat in.zson 
{_path:"conn",ts:1970-01-01T00:00:01Z,uid:"CBrzd94qfowOqJwCHa" (bstring)} (=0)

$ zq -z in.zson 
{_path:"conn",ts:1970-01-01T00:00:01Z,uid:"CBrzd94qfowOqJwCHa" (bstring)} (=0)

$ cat in1.zson 
{_path:"conn",ts:1970-01-01T00:00:01Z,uid:"CBrzd94qfowOqJwCHa" (bstring)} (=0

$ zq -z in1.zson 
[no output]

$ echo $?
0

$ cat in2.zson 
{_path:"conn",ts:1970-01-01T00:00:01Z,uid:"CBrzd94qfowOqJwCHa" (bstring) (=0)

$ zq -z in2.zson 
[no output]

$ echo $?
0
```

**Comment by philrz** [comment]
Here's another one from this test:

[https://github.com/brimsec/zq/blob/79b4821bf60de2c3671fdfd46e15b83898cbaa35/ppl/zqd/handlers_test.go#L456-L461](https://github.com/brimsec/zq/blob/79b4821bf60de2c3671fdfd46e15b83898cbaa35/ppl/zqd/handlers_test.go#L456-L461)

Manual run in its original TZNG form:

```
$ cat in.tzng 
#0:record[_path:string,ts:time,uid:bstring]
0:[conn;1;CBrzd94qfowOqJwCHa;]
detectablebutbadline
16:20:39-phil@PhilMac:~/work/newbug$ zq -z in.tzng 
in.tzng: line 3: bad format
```

Attempting a ZSON equivalent, the error message seems particularly specific, since nothing about this input screams "failed attempt at `enum`" to me.

```
$ cat in.zson 
{_path:"conn",ts:1970-01-01T00:00:01Z,uid:"CBrzd94qfowOqJwCHa" (bstring)} (=0)
detectablebutbadline

$ zq -z in.zson 
in.zson: identifier "detectablebutbadline" must be enum and requires decorator
```

**Comment by mccanne** [comment]
Yes, I agree this is an odd error message and arises from the fact that in the zson parser, the only possible value given the context is an enum, but then it doesn't parse as enum because it doesn't have the required type decorate.  We should fix this somehow...
